<h3>Name : Akinremi Bunmi Josephine.

<h3>kaggle_Id : bumie.

<h3>Student_Status: 100 level.

<h3>Gender : Female.

<h3>Special_Services : AI+ Club Campus Co-ordinator

<h3>Name_of_School : Obafemi Awolowo University

In [2]:
# Most Important
import pandas as pd
import numpy as np

# For plotting
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

<h3>Import the testdata, train data, submission file

In [3]:
test = pd.read_csv('test.csv')
train = pd.read_csv('train.csv')
sam = pd.read_csv('sample_submission2.csv')

<h3>Lets take a brief look at the train data

In [ ]:
train.head()

In [ ]:
train.shape

In [4]:
train.describe()

Trainings_Attended  Year_of_birth  Last_performance_score  \
count        38312.000000   38312.000000            38312.000000   
mean             2.253680    1986.209334                7.698959   
std              0.609443       7.646047                3.744135   
min              2.000000    1950.000000                0.000000   
25%              2.000000    1982.000000                5.000000   
50%              2.000000    1988.000000                7.500000   
75%              2.000000    1992.000000               10.000000   
max             11.000000    2001.000000               12.500000   

       Year_of_recruitment   Targets_met  Previous_Award  \
count         38312.000000  38312.000000    38312.000000   
mean           2013.139695      0.352996        0.023152   
std               4.261451      0.477908        0.150388   
min            1982.000000      0.000000        0.000000   
25%            2012.000000      0.000000        0.000000   
50%            2014.000000      0.000000        0.000000   
75%            2016.000000      1.000000        0.000000   
max            2018.000000      1.000000        1.000000   

       Training_score_average  Promoted_or_Not  
count            38312.000000     38312.000000  
mean                55.366465         0.084595  
std                 13.362741         0.278282  
min                 31.000000         0.000000  
25%                 43.000000         0.000000  
50%                 52.000000         0.000000  
75%                 68.000000         0.000000  
max                 91.000000         1.000000

<b> We can see that the train dataset has 38312 rows and 19 columns

<h3> Lets take a brief look at the test data

In [ ]:
test.head()

In [ ]:
test.shape

<b> We can see that the test dataset has 16496 columns and 18 rows</b>

<h3> Does our dataset contain any null value? Lets find out!

In [ ]:
train.info()

<h3> Hmmn, we can see that the column "Qualification" has only 36633 of its values that are not null. 
<h3> We can also see the various datatypes present in the dataset.(We'll deal with these later)

<h3>Lets divide the dataset into biased and unbiased. 
    

<h3>Lets first take a look at the unbiased dataset

<h4>The Unbiased dataset contains the following columns: Qualification, Target_met, Previous_Award, Trainning_Score_Average,
<h4>Foregin_Schooled, Past_Disiciplanry_Action, Previous_Intradepartmental_Movement

<h3> First of all, lets create a dataframe containing the test and train data

In [ ]:
itrain = train
itest = test

<h3>Lets Review the Qualification Column.

In [ ]:
train.Qualification.value_counts()

<h2>For better view let's plot these columns!

In [ ]:
def plot_graph(x, y):
    fig, ax = plt.subplots(figsize = (13, 4))
    sns.barplot(x, y)
    plt.xlabel('x', fontSize=13)
    plt.xticks(rotation='90')
    plt.ylabel('Promoted_or_Not', fontSize=13)
    plt.show()

In [ ]:
x = test.Qualification
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(18,8))
train['Qualification'].value_counts().plot.pie(explode=[0,0.1,0.2],autopct='%1.1f%%',ax=ax[0], shadow=True)
ax[0].set_title('Qualification vs promotion')
ax[0].set_ylabel('Qualification')
sns.factorplot('Qualification', 'Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Qualification vs promotion')
plt.show()

In [ ]:
<h4> From the above, we can say that more of the

Lets Review the Target_met column

In [ ]:
train.Targets_met.value_counts()

In [ ]:
sns.factorplot('Targets_met','Promoted_or_Not',hue='Targets_met',data=train)
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(18,8))
train[['Targets_met','Promoted_or_Not']].groupby(['Targets_met']).mean().plot.bar(ax=ax[0])
ax[0].set_title('Targets_met vs Promoted_or_Not')
sns.countplot('Targets_met',hue='Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Targets_met:Promoted or Not')
plt.show()

In [ ]:
x = test.Targets_met
y = itrain.Promoted_or_Not

plot_graph(x, y)

Lets review the Previous_Award column

In [ ]:
train.Previous_Award.value_counts()

Lets Review the Promoted_or_Not Column

In [ ]:
#For better view let's plot these percents!
x = test.Previous_Award
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Previous_Award','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Previous_Award vs Promoted_or_Not')
sns.factorplot('Previous_Award','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Previous_Award vs Promoted_or_Not')
plt.close(2)
plt.show()

In [ ]:
sns.factorplot('Previous_Award','Promoted_or_Not',hue='Previous_Award',data=train)
plt.show()

Lets review the Trainning score average column

In [ ]:
x = train.Training_score_average
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.factorplot('Training_score_average','Promoted_or_Not',data=train)
plt.show()

Lets review the Foregin_Schooled column

In [ ]:
train.Foreign_schooled.value_counts()

In [ ]:
x = test.Foreign_schooled
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.distplot(train[train['Foreign_schooled']=='Yes'].Promoted_or_Not,ax=ax[0])
ax[0].set_title('Schooled Aborad:Yes')
sns.distplot(train[train['Foreign_schooled']=='No'].Promoted_or_Not,ax=ax[1])
ax[1].set_title('Schooled Abroad:No')
plt.show()

Lets take a look at the Past_Disciplinary_Action column!,

And see those who has been disciplined(bad boys) before!

In [ ]:
train.Past_Disciplinary_Action.value_counts()

In [ ]:
sns.factorplot('Past_Disciplinary_Action','Promoted_or_Not',data=train)
plt.show()

In [ ]:
x = test.Past_Disciplinary_Action
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.distplot(train[train['Past_Disciplinary_Action']=='Yes'].Promoted_or_Not,ax=ax[0])
ax[0].set_title('Past_Disciplinary_Action:Yes')
sns.distplot(train[train['Past_Disciplinary_Action']=='No'].Promoted_or_Not,ax=ax[1])
ax[1].set_title('Past_Disciplinary_Action:No')
plt.show()

Lets take a look at the Previous_Intradepartmental_Movement column. Some might have even been promoted before!

In [ ]:
sns.factorplot('Previous_IntraDepartmental_Movement','Promoted_or_Not',data=train)
plt.show()

In [ ]:
x = itrain.Previous_IntraDepartmental_Movement
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(18,8))
train[['Previous_IntraDepartmental_Movement','Promoted_or_Not']].groupby(['Previous_IntraDepartmental_Movement']).mean().plot.bar(ax=ax[0])
ax[0].set_title('Previous_IntraDepartmental_Movement vs Promoted_or_Not')
sns.countplot('Previous_IntraDepartmental_Movement',hue='Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Previous_IntraDepartmental_Movement:Promoted or Not')
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,15))
sns.countplot('Previous_IntraDepartmental_Movement',data=train,ax=ax[0])
ax[0].set_title('Previous_IntraDepartmental_Movement')
sns.countplot('Previous_IntraDepartmental_Movement',hue='Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Previous_IntraDepartmental_Movement for Promoted_or_Not')
plt.subplots_adjust(wspace=0.2,hspace=0.5)
plt.show()

Now, after we are done with that, lets take a look at the Biased dataset.

<h4>The Biased dataset contains the following : </h4>
Division, Marital_Status, Gender, Channel_of_Recruitment, Year_of_birth, Trainings_Attended, Last_performance_score, No_of_previous employers

In [ ]:
x = train.Division
y = train.Promoted_or_Not
plot_graph(x, y)

In [ ]:
x = test.Division
y = train.Promoted_or_Not
plot_graph(x, y)

In [ ]:
sns.factorplot('Division', 'Promoted_or_Not',data=train)
plt.xlabel('Division Columns', fontSize=13)
plt.xticks(rotation='90')
plt.ylabel('Promoted_or_Not', fontSize=13)
plt.show()

In [ ]:
sns.barplot(y,x,orient = 'h')
plt.grid('on')
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(18,8))
train['Division'].value_counts().plot.pie(explode=[0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8],autopct='%1.1f%%',ax=ax[0], shadow=True)
ax[0].set_title('Division vs promotion')
ax[0].set_ylabel('Division')
sns.factorplot('Division', 'Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Division vs promotion')
plt.show()

<h3> From these, we can see that this column is biased as some divisions are more promoted than some

In [ ]:
itrain.Gender.value_counts()

In [ ]:
x = test.Gender
y = train.Promoted_or_Not

plot_graph(x, y)

In [ ]:
#For better view let's plot these percents!
x = train.Gender
y = train.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
train['Gender'].value_counts().plot.pie(explode=[0,0.1],autopct='%1.1f%%',ax=ax[0], shadow=True)
ax[0].set_title('Gender vs Promoted_or_Not')
sns.factorplot('Gender','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Gender vs Promoted_or_Not')
plt.close(2)
plt.show()

<h3> From here, we can see that there are more males than females but, more females are promoted than males!
    
<h3> Lets look at the Channel_of_Recruitment column

In [ ]:
train.Channel_of_Recruitment.value_counts()

In [ ]:
x = test.Channel_of_Recruitment
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
#For better view let's plot these percents!
x = train.Channel_of_Recruitment
y = train.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Channel_of_Recruitment','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Channel_of_Recruitment vs Promoted_or_Not')
sns.factorplot('Channel_of_Recruitment','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Channel_of_Recruitment vs Promoted_or_Not')
plt.close(2)
plt.show()

<h3> Am sure you'll agree that with me that those who had special referral even got promoted than those who cme through the normal qualified way
    
<h3> So now, Lets take a look at the Trainings_Attended column

In [ ]:
train.Trainings_Attended.value_counts()

In [ ]:
x = train.Trainings_Attended
y = train.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.barplot(y,x,orient = 'h')
plt.grid('on')
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Trainings_Attended','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Trainings_Attended vs Promoted_or_Not')
sns.factorplot('Trainings_Attended','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Trainings_Attended vs Promoted_or_Not')
plt.close(2)
plt.show()

<h3> Well, its just obivious that these employees were not promoted based on their performance

<h3> Without much ado, lets move to the Year_of_birth column

In [ ]:
train.Year_of_birth.value_counts()

In [ ]:
#For better view let's plot these percents!
x = train.Year_of_birth
y = train.Promoted_or_Not

plot_graph(x, y)

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Year_of_birth','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Year_of_birth vs Promoted_or_Not')
sns.factorplot('Year_of_birth','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Year_of_birth vs Promoted_or_Not')
plt.close(2)
plt.show()

In [ ]:
sns.barplot(y,x,orient = 'h')
plt.grid('on')
plt.show()

<h3> From these, we can really tell whose who because they are ll closely packed together

<h3> Now, moving on to Last_performance_score column

In [ ]:
train.Last_performance_score.value_counts()

In [ ]:
x = test.Last_performance_score
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
#For better view let's plot these percents!
x = itrain.Last_performance_score
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.barplot(y,x,orient = 'h')
plt.grid('on')
plt.show()

In [ ]:
'NaN' in itrain.Last_performance_score

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Last_performance_score','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Last_performance_score vs Promoted_or_Not')
sns.factorplot('Last_performance_score','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Last_performance_score vs Promoted_or_Not')
plt.close(2)
plt.show()

<h3> We can also see some imbalances especially with those who scored 0.0 points

<h3>Ok, lets look at the Year_of_recruitment column

In [ ]:
train.Year_of_recruitment.value_counts()

In [ ]:
x = test.Year_of_recruitment
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
x = itrain.Year_of_recruitment
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.factorplot('Year_of_recruitment','Promoted_or_Not',hue='Promoted_or_Not',data=train)
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(18,8))
train[['Year_of_recruitment','Promoted_or_Not']].groupby(['Year_of_recruitment']).mean().plot.bar(ax=ax[0])
ax[0].set_title('Promoted_or_Not vs Year_of_recruitment')
sns.countplot('Year_of_recruitment',hue='Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Year_of_recruitment:Promoted or Not')
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Year_of_recruitment','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Year_of_recruitment vs Promoted_or_Not')
sns.factorplot('Year_of_recruitment','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Year_of_recruitment vs Promoted_or_Not')
plt.close(2)
plt.show()

<h3> We can also see imbalances here too with this column
    
<h3> Next! Taking a look at the State_Of_Origin column,

In [ ]:
train.State_Of_Origin.value_counts()

In [ ]:
#For better view let's plot these percents!
x = itrain.State_Of_Origin
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
x = test.State_Of_Origin
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.factorplot('State_Of_Origin','Promoted_or_Not',hue='State_Of_Origin',data=train)
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('State_Of_Origin','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('State_Of_Origin vs Promoted_or_Not')
sns.factorplot('State_Of_Origin','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('State_Of_Origin vs Promoted_or_Not')
plt.close(2)
plt.show()

<h3> Also in this column, we can see various irregularities in the pattern


<h3> Now, lets look at the Foreign_schooled column

In [ ]:
train.Marital_Status.value_counts()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(20,8))
sns.barplot('Marital_Status','Promoted_or_Not',data=train,ax=ax[0])
ax[0].set_title('Marital_Status vs Promoted_or_Not')
sns.factorplot('Marital_Status','Promoted_or_Not',data=train,ax=ax[1])
ax[1].set_title('Marital_Status vs Promoted_or_Not')
plt.close(2)
plt.show()

In [ ]:
x = test.Marital_Status
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.factorplot('Marital_Status','Promoted_or_Not',hue='Marital_Status',data=train)
plt.show()

In [ ]:
x = train.Marital_Status
y = train.Promoted_or_Not

plot_graph(x, y)

In [ ]:
train.No_of_previous_employers.value_counts()

In [ ]:
#For better view let's plot these percents!
x = itrain.No_of_previous_employers
y = itrain.Promoted_or_Not

plot_graph(x, y)

In [ ]:
sns.factorplot('No_of_previous_employers','Promoted_or_Not',hue='No_of_previous_employers',data=train)
plt.show()

In [ ]:
f,ax=plt.subplots(1,2,figsize=(18,8))
itrain['No_of_previous_employers'].value_counts().plot.pie(explode=[0,0.1,0.2,0.3,0.4,0.5,0.6],autopct='%1.1f%%',ax=ax[0], shadow=True)
ax[0].set_title('No_of_previous_employers vs promotion')
ax[0].set_ylabel('No_of_previous_employers')
sns.factorplot('No_of_previous_employers', 'Promoted_or_Not',data=itrain,ax=ax[1])
ax[1].set_title('No_of_previous_employers vs promotion')
plt.show()

<h3> Here also, we can basically see the irregularity in thr promoting pattern
    
<h3> So now, we are basically done with visualization

<h1>Now, lets Label Encode and One Hot Encode our data, where neccesary and drop few unimportant columns

In [3]:
itrain = pd.get_dummies(train, columns = ['Channel_of_Recruitment','Marital_Status', 'Division', 'Previous_IntraDepartmental_Movement', 'State_Of_Origin', 'Gender'])
itest = pd.get_dummies(test, columns = ['Channel_of_Recruitment', 'Marital_Status', 'Division', 'Previous_IntraDepartmental_Movement', 'State_Of_Origin', 'Gender'])

In [4]:
itrain['Qualification'][itrain['Qualification'] == 'MSc, MBA and PhD'] = 2
itrain['Qualification'][itrain['Qualification'] == 'First Degree or HND'] = 0
itrain['Qualification'][itrain['Qualification'] == 'Non-University Education'] = 1


itest['Qualification'][itest['Qualification'] == 'MSc, MBA and PhD'] = 2
itest['Qualification'][itest['Qualification'] == 'First Degree or HND'] = 0
itest['Qualification'][itest['Qualification'] == 'Non-University Education'] = 1

In [5]:
itrain.Marital_Status.value_counts()

AttributeError: 'DataFrame' object has no attribute 'Marital_Status'

In [13]:
itrain['Marital_Status'][itrain['Marital_Status'] == 'MSc, MBA and PhD'] = 2
itrain['Marital_Status'][itrain['Marital_Status'] == 'First Degree or HND'] = 0
itrain['Marital_Status'][itrain['Marital_Status'] == 'Non-University Education'] = 1


itest['Marital_Status'][itest['Marital_Status'] == 'MSc, MBA and PhD'] = 2
itest['Marital_Status'][itest['Marital_Status'] == 'First Degree or HND'] = 0
itest['Marital_Status'][itest['Marital_Status'] == 'Non-University Education'] = 1

KeyError: 'Marital_Status'

In [14]:
itrain['Targets_met'][itrain['Targets_met'] == 1] = 0
itrain['Targets_met'][itrain['Targets_met'] == 0] = 1

itest['Targets_met'][itest['Targets_met'] == 1] = 0
itest['Targets_met'][itest['Targets_met'] == 0] = 1

In [15]:
itrain['Previous_Award'][itrain['Previous_Award'] == 1] = 0
itrain['Previous_Award'][itrain['Previous_Award'] == 0] = 1

itest['Previous_Award'][itest['Previous_Award'] == 1] = 0
itest['Previous_Award'][itest['Previous_Award'] == 0] = 1

In [16]:
itrain.Foreign_schooled[itrain.Foreign_schooled == 'Yes'] = 0
itrain.Foreign_schooled[itrain.Foreign_schooled == 'No'] = 1

itest.Foreign_schooled[itest.Foreign_schooled == 'Yes'] = 0
itest.Foreign_schooled[itest.Foreign_schooled == 'No'] = 1

In [17]:
itrain.Past_Disciplinary_Action[itrain.Past_Disciplinary_Action == 'No'] = 0
itrain.Past_Disciplinary_Action[itrain.Past_Disciplinary_Action == 'Yes'] = 1

itest.Past_Disciplinary_Action[itest.Past_Disciplinary_Action == 'No'] = 0
itest.Past_Disciplinary_Action[itest.Past_Disciplinary_Action == 'Yes'] = 1

In [18]:
itrain['Last_performance_score'][itrain['Last_performance_score'] == 12.5] = 0
itrain['Last_performance_score'][itrain['Last_performance_score'] == 10.0] = 1
itrain['Last_performance_score'][itrain['Last_performance_score'] == 7.5] = 2
itrain['Last_performance_score'][itrain['Last_performance_score'] == 5.0] = 3
itrain['Last_performance_score'][itrain['Last_performance_score'] == 2.5] = 4
itrain['Last_performance_score'][itrain['Last_performance_score'] == 0.0] = 5


itest['Last_performance_score'][itest['Last_performance_score'] == 12.5] = 0
itest['Last_performance_score'][itest['Last_performance_score'] == 10.0] = 1
itest['Last_performance_score'][itest['Last_performance_score'] == 7.5] = 2
itest['Last_performance_score'][itest['Last_performance_score'] == 5.0] = 3
itest['Last_performance_score'][itest['Last_performance_score'] == 2.5] = 4
itest['Last_performance_score'][itest['Last_performance_score'] == 0.0] = 5

In [6]:
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 11] = 0
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 10] = 1
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 9] = 2
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 8] = 3
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 7] = 4
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 6] = 5
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 5] = 6
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 4] = 7
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 3] = 8
itrain['Trainings_Attended'][itrain['Trainings_Attended'] == 2] = 9


itest['Trainings_Attended'][itest['Trainings_Attended'] == 11] = 0
itest['Trainings_Attended'][itest['Trainings_Attended'] == 9] = 1
itest['Trainings_Attended'][itest['Trainings_Attended'] == 8] = 2
itest['Trainings_Attended'][itest['Trainings_Attended'] == 7] = 3
itest['Trainings_Attended'][itest['Trainings_Attended'] == 6] = 4
itest['Trainings_Attended'][itest['Trainings_Attended'] == 5] = 5
itest['Trainings_Attended'][itest['Trainings_Attended'] == 4] = 6
itest['Trainings_Attended'][itest['Trainings_Attended'] == 3] = 7
itest['Trainings_Attended'][itest['Trainings_Attended'] == 2] = 8

In [8]:
y = itest[itest.columns]
z = itrain[itrain.columns]

In [9]:
# to write a function that compares the itest and itrain
# and checks for columns present in the itrain but absent in the itest
for i in z:
    if i in y:
        pass
    else:
        print(i)

Promoted_or_Not
Year_of_birth_1950
Year_of_birth_1952
Year_of_birth_1955
Year_of_birth_1956
Year_of_birth_1957


In [7]:
itrain.drop([, 'Year_of_birth'], axis=1, inplace=True)
itest.drop([, 'Year_of_birth'], axis=1, inplace=True)

In [10]:
itrain.drop(['Year_of_birth_1950', 'Year_of_birth_1952', 'Year_of_birth_1955', 'Year_of_birth_1956', 'Year_of_birth_1957'], axis=1, inplace = True)

In [11]:
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 'More than 5'] = 0
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 5] = 1
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 4] = 2
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 3] = 3
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 2] = 4
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 1] = 5
itrain['No_of_previous_employers'][itrain['No_of_previous_employers'] == 0] = 6

itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 'More than 5'] = 0
itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 5] = 1
itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 4] = 2
itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 3] = 3
itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 2] = 4
itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 1] = 5
itest['No_of_previous_employers'][itest['No_of_previous_employers'] == 0] = 6

<h3>Some of these columns are not integers. Some are objects, category, etc.
Lets convert it all to numeric form

In [19]:
itrain.Qualification = pd.to_numeric(itrain.Qualification)
itrain.Foreign_schooled = pd.to_numeric(itrain.Foreign_schooled)
itrain.Training_score_average = pd.to_numeric(itrain.Training_score_average)
itrain.Past_Disciplinary_Action = pd.to_numeric(itrain.Past_Disciplinary_Action)
itrain.No_of_previous_employers = pd.to_numeric(itrain.No_of_previous_employers)
itrain.Last_performance_score = pd.to_numeric(itrain.Last_performance_score)

itest.Qualification = pd.to_numeric(itest.Qualification)
itest.Foreign_schooled = pd.to_numeric(itest.Foreign_schooled)
itest.Past_Disciplinary_Action = pd.to_numeric(itest.Past_Disciplinary_Action)
itest.No_of_previous_employers = pd.to_numeric(itest.No_of_previous_employers)
itest.Training_score_average = pd.to_numeric(itest.Training_score_average)
itest.Last_performance_score = pd.to_numeric(itest.Last_performance_score)

In [20]:
itrain.shape, itest.shape

((38312, 111), (16496, 110))

<h3>Some of these columns contain null values. 
Lets fill them up with their mode

In [21]:
for i in itrain.columns:
    itrain[i].fillna(itrain[i].mode()[0], inplace = True)
    
for i in itest.columns:
    itest[i].fillna(itest[i].mode()[0], inplace = True)

<h3> Lets check to see if thats all done

In [ ]:
itrain.info()

In [ ]:
itest.info()

In [22]:
# lets assign a variable to EmployeeNo and drop it
target_Id = itest.EmployeeNo
itrain.drop(['EmployeeNo'], axis=1, inplace=True)
itest.drop(['EmployeeNo'], axis=1, inplace=True)

In [23]:
# lets create a variable for the column we are to predict and drop it
target = itrain.Promoted_or_Not
itrain.drop(['Promoted_or_Not'], axis=1, inplace=True)

In [ ]:
itrain.info()

In [ ]:
itest.info()

<h3> Since thats all done, Lets move to the next stage!

<h1> Making Predictions Using Machine Models

In [ ]:
# Since the data is highly imbalanced, we can upsample the column
# Creating a function "using_smote"
def using_smote(X, y):
    from imblearn.over_sampling import SMOTE
    sm = SMOTE()
    X, y = sm.fit_sample(X, y)
    return X, y
# Using the fuction on the itrain and target 
train_val, target_val = using_smote(itrain, target)

In [ ]:
# Using the fuction on the itrain and target 
train_val, target_val = using_smote(itrain, target)

In [24]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train,y_test= train_test_split(itrain, target , test_size = 0.22, random_state = 42)

In [25]:
itrain.shape, itest.shape

((38312, 109), (16496, 109))

In [26]:
# import the machine models to be used
from sklearn.ensemble import RandomForestClassifier #Random Forest
from sklearn.ensemble import GradientBoostingClassifier

# import the tools for metrics
from sklearn.model_selection import train_test_split #training and testing data split
from sklearn import metrics #accuracy measure
from sklearn.metrics import accuracy_score, precision_score, f1_score

<h1>* Using Random Forest Classifier

In [29]:
model = RandomForestClassifier(n_estimators=270, random_state=3, max_features='auto', n_jobs=4, bootstrap=True, class_weight=None, criterion='gini')
model.fit(x_train, y_train)
model.score(x_train, y_train),model.score(x_test, y_test)

(0.9994980423652243, 0.9298849211057065)

In [30]:
# predict using the X_test
pred1 = model.predict(x_test)
# Get the f1 score
print('The f1_score is', f1_score(pred1, y_test))

The f1_score is 0.3322033898305085


In [31]:
# see the overall performance of the data
from sklearn.metrics import classification_report
print('çlassification_report')
print(classification_report(pred1, y_test))

çlassification_report
              precision    recall  f1-score   support

           0       1.00      0.93      0.96      8264
           1       0.20      0.89      0.33       165

   micro avg       0.93      0.93      0.93      8429
   macro avg       0.60      0.91      0.65      8429
weighted avg       0.98      0.93      0.95      8429



In [32]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, pred1))

[[7691   18]
 [ 573  147]]


In [ ]:
# predict for the itest data 
pred1a = model.predict(itest)

In [ ]:
output = pd.DataFrame(data = {'EmployeeNo':target_Id, 'Promoted_or_Not':pred1a})
output.to_csv(path_or_buf = 'rand-6bumiepredi.csv', index = False, quoting = 3)

<h1> Using the Gradient Boosting Classifier

In [34]:
model = GradientBoostingClassifier(loss='exponential', learning_rate = 0.2,subsample = 0.7, n_estimators = 300, random_state=5, max_features='auto',
                                   max_depth=4, warm_start=True, min_samples_leaf=5, min_samples_split=4, )
model.fit(x_train, y_train) 
print('Accuracy score for x_train is ', model.score(x_train, y_train))
print('Accuracy score for x_test is ', model.score(x_test, y_test))

      Iter       Train Loss      OOB Improve   Remaining Time 
         1           0.5342           0.0172            2.74m
         2           0.5228           0.0126            2.64m
         3           0.5179           0.0090            2.58m
         4           0.5006           0.0104            2.63m
         5           0.5031           0.0061            2.54m
         6           0.4960           0.0039            2.63m
         7           0.4825           0.0091            2.58m
         8           0.4818           0.0056            2.72m
         9           0.4782           0.0022            2.69m
        10           0.4727           0.0021            2.68m
        11           0.4692           0.0020            2.70m
        12           0.4716           0.0035            2.69m
        13           0.4657           0.0019            2.67m
        14           0.4649           0.0009            2.65m
        15           0.4604           0.0036            2.63m
       

       133           0.3897          -0.0008            1.75m
       134           0.3826          -0.0003            1.74m
       135           0.3823          -0.0005            1.73m
       136           0.3860          -0.0005            1.72m
       137           0.3854          -0.0008            1.71m
       138           0.3859          -0.0003            1.70m
       139           0.3851          -0.0006            1.69m
       140           0.3800          -0.0003            1.68m
       141           0.3804          -0.0000            1.67m
       142           0.3800          -0.0005            1.66m
       143           0.3836          -0.0001            1.64m
       144           0.3822          -0.0007            1.63m
       145           0.3838          -0.0006            1.62m
       146           0.3809          -0.0006            1.62m
       147           0.3781          -0.0005            1.61m
       148           0.3766          -0.0004            1.60m
       1

       266           0.3466          -0.0004           22.62s
       267           0.3501          -0.0004           21.96s
       268           0.3466          -0.0004           21.31s
       269           0.3506          -0.0002           20.63s
       270           0.3490          -0.0005           19.95s
       271           0.3455          -0.0004           19.28s
       272           0.3518          -0.0006           18.60s
       273           0.3403          -0.0004           17.92s
       274           0.3477          -0.0002           17.25s
       275           0.3476          -0.0002           16.57s
       276           0.3493          -0.0008           15.90s
       277           0.3416          -0.0005           15.23s
       278           0.3461          -0.0005           14.56s
       279           0.3436          -0.0004           13.88s
       280           0.3440          -0.0003           13.22s
       281           0.3465          -0.0005           12.55s
       2

In [48]:
# predict using the x_test
pred2 = model.predict(x_test)
# predict using the itest data
pred2a = model.predict(itest)

In [36]:
# get the f1 score
print('The f1_score is',f1_score(pred2, y_test))

The f1_score is 0.4585062240663901


In [37]:
# see the overall performance of the data
from sklearn.metrics import classification_report
print('çlassification_report')
print(classification_report(pred2, y_test))

çlassification_report
              precision    recall  f1-score   support

           0       1.00      0.94      0.97      8185
           1       0.31      0.91      0.46       244

   micro avg       0.94      0.94      0.94      8429
   macro avg       0.65      0.92      0.71      8429
weighted avg       0.98      0.94      0.95      8429



In [38]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, pred2))

[[7686   23]
 [ 499  221]]


In [ ]:
# Lets Export It To A Csv File
output = pd.DataFrame(data = {'EmployeeNo':target_Id, 'Promoted_or_Not':pred2a})
output.to_csv(path_or_buf = 'gb-4bumiepredi.csv', index = False, quoting = 3)

<h1>* Using XGBoost Classifier

In [39]:
from xgboost import XGBClassifier
xgb = XGBClassifier(learning_rate = 0.2, num_boosting_rounds = 270 , max_depth=10, subsample=0.8)
xgb.fit(x_train, y_train)
print('The Accuracy Score For x_test Is : ', xgb.score(x_train, y_train)), print('The Accuracy Score For x_test Is : ', xgb.score(x_test, y_test))

The Accuracy Score For x_test Is :  0.9522805608539973
The Accuracy Score For x_test Is :  0.9378336694744335


(None, None)

<h3> Lets convert the itest data to an array first

In [40]:
itest = np.array(itest)

In [41]:
col = [x for x in itrain.columns]

itest = pd.DataFrame(itest, columns = col)

In [42]:
pred3 = xgb.predict(x_test)
pred3a = xgb.predict(itest)
f1_score(pred3, y_test)

0.4575569358178053

In [43]:
# see the overall performance of the data
from sklearn.metrics import classification_report
print('çlassification_report')
print(classification_report(pred3, y_test))

çlassification_report
              precision    recall  f1-score   support

           0       1.00      0.94      0.97      8183
           1       0.31      0.90      0.46       246

   micro avg       0.94      0.94      0.94      8429
   macro avg       0.65      0.92      0.71      8429
weighted avg       0.98      0.94      0.95      8429



In [46]:
output = pd.DataFrame(data = {'EmployeeNo':target_Id, 'Promoted_or_Not':pred3a})
output.to_csv(path_or_buf = 'xgb-3bumiepredi.csv', index = False, quoting = 3)

<h3> Lets try stacking models together to see if we would have a better precision and recall score

In [47]:
from vecstack import stacking
from xgboost import XGBClassifier
models = [RandomForestClassifier(n_estimators=270, random_state=3, max_features='auto', n_jobs=2, bootstrap=True, class_weight=None, criterion='gini'), GradientBoostingClassifier(learning_rate = 0.2,subsample=0.8, n_estimators = 270, random_state=5, max_features='auto', max_depth=10), XGBClassifier(learning_rate = 0.2, num_boosting_rounds = 270 , max_depth=10, subsample=0.8)]
a_train, a_test = stacking(models,                   
                           x_train, y_train, x_test,   
                           regression=False, 
     
                           mode='oof_pred_bag', 
       
                           needs_proba=False,
         
                           save_dir=None, 
            
                           metric=f1_score, 
    
                           n_folds=5, 
                 
                           stratified=True,
            
                           shuffle=True,  
            
                           random_state=0,    
         
                           verbose=2)

task:         [classification]
n_classes:    [2]
metric:       [f1_score]
mode:         [oof_pred_bag]
n_models:     [3]

model  0:     [RandomForestClassifier]
    fold  0:  [0.31699346]
    fold  1:  [0.37558685]
    fold  2:  [0.33118971]
    fold  3:  [0.38425197]
    fold  4:  [0.38184664]
    ----
    MEAN:     [0.35797373] + [0.02816895]
    FULL:     [0.35843661]

model  1:     [GradientBoostingClassifier]
    fold  0:  [0.43193717]
    fold  1:  [0.44955300]
    fold  2:  [0.44359949]
    fold  3:  [0.45959596]
    fold  4:  [0.46033810]
    ----
    MEAN:     [0.44900475] + [0.01059659]
    FULL:     [0.44906338]

model  2:     [XGBClassifier]
    fold  0:  [0.46902655]
    fold  1:  [0.49928264]
    fold  2:  [0.50578035]
    fold  3:  [0.49784173]
    fold  4:  [0.48695652]
    ----
    MEAN:     [0.49177756] + [0.01288433]
    FULL:     [0.49188876]



In [ ]:
a_train

<h3> Lets save the predictions from random forest, gradient boosting and xgboost classification as a data frame

In [ ]:
new_test =  pd.DataFrame(pred1a)
array = pred2a
array1 = pred3a
new_test[1] = array
new_test[2] = array1
new_test = np.array(new_test)

In [ ]:
b_train['1'].value_counts()

<h3> Lets now test the new test and the new train(a_train, a_test)

In [ ]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(a_train, y_train)
lr.score(a_train, y_train), lr.score(a_test, y_test)

In [ ]:
# Lets test it on the x_test data
pred = lr.predict(a_test)
# predict using the itest data
pred3a = lr.predict(b_train)

In [ ]:
# get the f1 score
f1_score(pred, y_test)

In [ ]:
# see the overall performance of the data
from sklearn.metrics import classification_report
print('çlassification_report')
print(classification_report(pred, y_test))

In [ ]:
output = pd.DataFrame(data = {'EmployeeNo':target_Id, 'Promoted_or_Not':pred3a})
output.to_csv(path_or_buf = 'lr-1bumiepredi.csv', index = False, quoting = 3)

In [ ]:
from sklearn import svm #support vector Machine
model=svm.SVC(kernel='',C=1,gamma=0.1) 
model.fit(a_train, y_train)
model.score(a_train, y_train), model.score(a_test, y_test)

In [ ]:
# Lets test it on the x_test data
pred = model.predict(a_test)
# predict using the itest data
pred3a = model.predict(b_train)

In [ ]:
# get the f1 score
f1_score(pred, y_test)

In [ ]:
# see the overall performance of the data
from sklearn.metrics import classification_report
print('çlassification_report')
print(classification_report(pred, y_test))

In [ ]:
output = pd.DataFrame(data = {'EmployeeNo':target_Id, 'Promoted_or_Not':pred3a})
output.to_csv(path_or_buf = 'svc-1bumiepredi.csv', index = False, quoting = 3)